server/src/main/java/org/elasticsearch/health/node/DiskHealthIndicatorService.java

```diff
@@ -0,0 +1,448 @@
+                blockedIndices.size(),
+                blockedIndices.size() == 1 ? "index is" : "indices are"
+            );
+        } else {
+            symptom = String.format(
+                Locale.ROOT,
+                "%d node%s with roles: [%s] %s out of disk or running low on disk space.",
+                allUnhealthyNodes.size(),
+                allUnhealthyNodes.size() == 1 ? "" : "s",
+                allRolesOnUnhealthyNodes.stream().sorted().collect(Collectors.joining(", ")),
+                allUnhealthyNodes.size() == 1 ? "is" : "are"
+            );
+        }
+        return symptom;
+    }
+
+    private List<HealthIndicatorImpact> getImpacts(
+        Set<String> indicesWithBlock,
+        Set<String> indicesOnRedNodes,
+        Set<String> indicesOnYellowNodes,
+        Set<String> nodesWithBlockedIndices,
+        Set<String> redDataNodes,
+        Set<String> yellowDataNodes,
+        Set<String> redMasterNodes,
+        Set<String> yellowMasterNodes,
+        Set<String> redNonDataNonMasterNodes,
+        Set<String> yellowNonDataNonMasterNodes
+    ) {
+        List<HealthIndicatorImpact> impacts = new ArrayList<>();
+        if (indicesWithBlock.isEmpty() == false
+            || indicesOnRedNodes.isEmpty() == false
+            || nodesWithBlockedIndices.isEmpty() == false
+            || redDataNodes.isEmpty() == false) {
+            impacts.add(
+                new HealthIndicatorImpact(1, "Cannot insert or update documents in the affected indices.", List.of(ImpactArea.INGEST))
+            );
+        } else if (indicesOnYellowNodes.isEmpty() == false || yellowDataNodes.isEmpty() == false) {
+            impacts.add(
+                new HealthIndicatorImpact(
+                    1,
+                    "At risk of not being able to insert or update documents in the affected indices.",
+                    List.of(ImpactArea.INGEST)
+                )
+            );
+        }
+        if (redMasterNodes.isEmpty() == false || yellowMasterNodes.isEmpty() == false) {
+            impacts.add(new HealthIndicatorImpact(2, "Cluster stability might be impaired.", List.of(ImpactArea.DEPLOYMENT_MANAGEMENT)));
+        }
+        if (redNonDataNonMasterNodes.isEmpty() == false || yellowNonDataNonMasterNodes.isEmpty() == false) {
+            impacts.add(
+                new HealthIndicatorImpact(2, "Some cluster functionality might be unavailable.", List.of(ImpactArea.DEPLOYMENT_MANAGEMENT))
+            );
+        }
+        return impacts;
+    }
+
+    private List<Diagnosis> getDiagnoses(
+        Set<String> indicesWithBlock,
+        Set<String> indicesOnRedNodes,
+        Set<String> indicesOnYellowNodes,
+        Set<String> nodesWithBlockedIndices,
+        Set<String> redDataNodes,
+        Set<String> yellowDataNodes,
+        Set<String> redMasterNodes,
+        Set<String> yellowMasterNodes,
+        Set<String> redNonDataNonMasterNodes,
+        Set<String> yellowNonDataNonMasterNodes
```

Comment:
I think we should group these a in a couple of records (similar to what we do in https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java#L349 )

Let's create an issue and refactor in a subsequent PR.